src/Parser/Hunter/Shared/Modules/Features/FocusUsage.js

```diff
@@ -0,0 +1,323 @@
+        }}
+        options={{
+          legend: {
+            display: false,
+          },
+          tooltips: {
+            bodyFontSize: 8,
+          },
+          cutoutPercentage: 45,
+          animation: false,
+          responsive: false,
+        }}
+        width={CHART_SIZE}
+        height={CHART_SIZE}
+      />
+    );
+  }
+
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    if (this.combatants.selected.spec === SPECS.MARKSMANSHIP_HUNTER) {
+      if (LIST_OF_MM_FOCUS_SPENDERS.every(id => spellId !== id)) {
+        return;
+      }
+      this.focusSpenderCastsMM[spellId].casts += 1;
+      this.focusSpenderCastsMM[spellId].focusUsed += event.classResources[0]['cost'] || 0;
+    } else if (this.combatants.selected.spec === SPECS.BEAST_MASTERY_HUNTER) {
+      if (LIST_OF_BM_FOCUS_SPENDERS.every(id => spellId !== id)) {
+        return;
+      }
+      this.focusSpenderCastsBM[spellId].casts += 1;
+      this.focusSpenderCastsBM[spellId].focusUsed += event.classResources[0]['cost'] || 0;
+    }
+  }
+  on_byPlayer_damage(event) {
+    const spellId = event.ability.guid;
+    if (spellId !== SPELLS.VOLLEY_ACTIVATED.id) {
+      return;
+    }
+    if (event.timestamp > (this.lastVolleyHit + BUFFER_MS)) {
+      if (this.combatants.selected.spec === SPECS.MARKSMANSHIP_HUNTER) {
+        this.focusSpenderCastsMM[spellId].casts += 1;
+        this.focusSpenderCastsMM[spellId].focusUsed += VOLLEY_FOCUS_COST;
+
+      }
+      if (this.combatants.selected.spec === SPECS.BEAST_MASTERY_HUNTER) {
+        this.focusSpenderCastsBM[spellId].casts += 1;
+        this.focusSpenderCastsBM[spellId].focusUsed += VOLLEY_FOCUS_COST;
+      }
+      this.lastVolleyHit = event.timestamp;
+    }
+  }
+
+  focusUsageChart() {
+    let totalFocusUsed = 0;
+    const items = [];
+    const bmFocusSpenders = [[SPELLS.COBRA_SHOT.id], [SPELLS.MULTISHOT.id], [SPELLS.KILL_COMMAND.id], [SPELLS.REVIVE_PET_AND_MEND_PET.id], [SPELLS.A_MURDER_OF_CROWS_TALENT_SHARED.id], [SPELLS.BARRAGE_TALENT.id], [SPELLS.VOLLEY_ACTIVATED.id]];
+    const mmFocusSpenders = [[SPELLS.AIMED_SHOT.id], [SPELLS.MARKED_SHOT.id], [SPELLS.REVIVE_PET_AND_MEND_PET.id], [SPELLS.BARRAGE_TALENT.id], [SPELLS.VOLLEY_ACTIVATED.id], [SPELLS.WINDBURST.id], [SPELLS.PIERCING_SHOT_TALENT.id], [SPELLS.BLACK_ARROW_TALENT.id], [SPELLS.EXPLOSIVE_SHOT_TALENT.id]];
+    if (this.combatants.selected.spec === SPECS.MARKSMANSHIP_HUNTER) {
+      mmFocusSpenders.forEach(focusSpender => {
+        if (this.focusSpenderCastsMM[focusSpender].casts > 0) {
+          items.push({
+            color: this.focusSpenderCastsMM[focusSpender].color,
+            label: this.focusSpenderCastsMM[focusSpender].name,
+            spellId: focusSpender,
+            value: this.focusSpenderCastsMM[focusSpender].focusUsed,
+          });
+        }
+      });
+      mmFocusSpenders.map(ability => totalFocusUsed += this.focusSpenderCastsMM[ability].focusUsed);
```

Comment:
This should be a reducer

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce